File: packages/format/src/digest.js
import { positionPhrase, readingOrder } from './geometry.js';
import { plainText, countWords, headingLevel } from './mdblocks.js';
import { usedRange, summarizeColumns, summaryScope } from './grid.js';
import { parseChartBlock, resolveChartSpec, describeChart, chartToMarkdownTable } from './chart.js';
import { indexToCol, toRef, displayValue, recalcSheet } from '@ai-studio/formula';

const KIND_LABEL = {
  text: '텍스트',
  image: '이미지',
  shape: '도형',
  table: '표',
  chart: '차트',
};

/**
 * Build `AI.md` — a single markdown file that flattens a whole project for LLM
 * consumption.
 *
 * The design bet: a model reasons far better about "상단 중앙, 전체 폭" than about
 * `x=96 y=220 w=1088`, and better about a sheet's column summary than about 500
 * rows of numbers. So the digest is not a dump of the source files; it is a
 * translation of them into prose a model already understands, with enough
 * structure (slide numbers, cell addresses, heading levels) to stay verifiable
 * against the source.
 */
export function buildDigest(project) {
  switch (project.type) {
    case 'deck':
      return deckDigest(project);
    case 'doc':
      return docDigest(project);
    case 'grid':
      return gridDigest(project);
    default:
      return `# ${project.manifest?.title ?? 'Untitled'}\n`;
  }
}

function header(project, subtitle) {
  const m = project.manifest ?? {};
  const modified = m.modified ? new Date(m.modified).toISOString().slice(0, 10) : '—';
  return [
    `# ${m.title ?? '제목 없음'}`,
    '',
    `> ${subtitle} · 최종 수정 ${modified}`,
    '',
    '<!-- 이 파일은 저장할 때마다 자동 생성됩니다. 직접 편집하면 다음 저장 시 덮어써집니다. -->',
    '',
  ];
}

/* -------------------------------------------------------------------- deck */

function deckDigest(project) {
  const slides = project.slides ?? [];
  const lines = header(project, `AI Studio 프레젠테이션 · 슬라이드 ${slides.length}장`);

  lines.push('## 목차', '');
  slides.forEach((slide, i) => {
    lines.push(`${i + 1}. ${slide.title || '(제목 없음)'} — \`${slide.layoutName}\``);
  });
  lines.push('');
  lines.push('---', '');

  slides.forEach((slide, i) => {
    lines.push(`## 슬라이드 ${i + 1} — ${slide.title || '(제목 없음)'}`);
    lines.push('');
    lines.push(`- 레이아웃: \`${slide.layoutName}\``);
    lines.push(`- 요소 ${slide.blocks.length}개`);
    if (slide.notes?.trim()) {
      lines.push(`- 발표자 노트: ${collapse(slide.notes)}`);
    }
    lines.push('');

    const ordered = readingOrder(slide.blocks);
    if (!ordered.length) {
      lines.push('_빈 슬라이드_', '');
      return;
    }

    ordered.forEach((block, bi) => {
      const where = positionPhrase(block, slide.canvas);
      const kind = KIND_LABEL[block.kind] ?? block.kind;
      lines.push(`### ${bi + 1}) ${kind} — ${where}`);
      lines.push('');
      const content = String(block.md ?? '').trim();

      // A chart is unreadable as a fenced JSON blob; give the model the shape and
      // the numbers instead.
      const chart = block.kind === 'chart' ? parseChartBlock(content) : null;
      if (chart) {
        lines.push(describeChart(chart));
        lines.push('');
        lines.push(chartToMarkdownTable(chart));
      } else {
        lines.push(content || '_(빈 요소)_');
      }
      lines.push('');
    });

    lines.push('---', '');
  });

  const allText = slides
    .flatMap((s) => s.blocks.map((b) => plainText(b.md)))
    .filter(Boolean)
    .join(' ');
  lines.push('## 전체 통계', '');
  lines.push(`- 슬라이드: ${slides.length}장`);
  lines.push(`- 요소: ${slides.reduce((a, s) => a + s.blocks.length, 0)}개`);
  lines.push(`- 본문 단어: 약 ${countWords(allText)}개`);
  lines.push(`- 발표자 노트가 있는 슬라이드: ${slides.filter((s) => s.notes?.trim()).length}장`);

  return `${lines.join('\n')}\n`;
}

/* --------------------------------------------------------------------- doc */

function docDigest(project) {
  const sections = project.sections ?? [];
  const lines = header(project, `AI Studio 문서 · 섹션 ${sections.length}개`);

  const outline = [];
  sections.forEach((section, si) => {
    section.blocks.forEach((b) => {
      const level = headingLevel(b.md);
      if (level) outline.push({ level, text: plainText(b.md), section: si + 1 });
    });
  });

  if (outline.length) {
    lines.push('## 목차', '');
    for (const item of outline) {
      lines.push(`${'  '.repeat(Math.max(0, item.level - 1))}- ${item.text}`);
    }
    lines.push('');
    lines.push('---', '');
  }

  sections.forEach((section, si) => {
    lines.push(`## 섹션 ${si + 1} — ${section.name}`);
    lines.push('');
    for (const block of section.blocks) {
      const md = String(block.md ?? '').trim();
      if (!md) continue;

      const chart = parseChartBlock(md);
      if (chart) {
        lines.push(describeChart(chart));
        lines.push('');
        lines.push(chartToMarkdownTable(chart));
      } else {
        lines.push(md);
      }
      if (block.override) {
        lines.push(`<!-- 서식: ${describeOverride(block.override)} -->`);
      }
      lines.push('');
    }
    lines.push('---', '');
  });

  const text = sections.flatMap((s) => s.blocks.map((b) => plainText(b.md))).join('\n');
  lines.push('## 전체 통계', '');
  lines.push(`- 섹션: ${sections.length}개`);
  lines.push(`- 단어: 약 ${countWords(text)}개`);
  lines.push(`- 글자(공백 제외): ${text.replace(/\s/g, '').length}자`);
  lines.push(`- 제목: ${outline.length}개`);

  return `${lines.join('\n')}\n`;
}

function describeOverride(o) {
  const parts = [];
  if (o.align) parts.push(`정렬 ${o.align}`);
  if (o.indent) parts.push(`들여쓰기 ${o.indent}px`);
  if (o.style?.italic) parts.push('기울임');
  if (o.style?.bold) parts.push('굵게');
  if (o.style?.color) parts.push(`색 ${o.style.color}`);
  if (o.style?.fontSize) parts.push(`크기 ${o.style.fontSize}px`);
  return parts.join(', ') || '기본';
}

/* -------------------------------------------------------------------- grid */

const GRID_DIGEST_ROWS = 60;

function gridDigest(project) {
  // Recalculate here rather than trusting the caller's cached `v` values: the
  // digest is generated on save, and a caller may hand us cells whose formulas
  // have never been evaluated (an API client, or a freshly parsed file).
  const sheets = (project.sheets ?? []).map((sheet) => ({
    ...sheet,
    cells: recalcSheet({ cells: sheet.cells, names: sheet.names }).cells,
  }));
  const lines = header(project, `AI Studio 스프레드시트 · 시트 ${sheets.length}개`);

  lines.push('## 시트 목록', '');
  sheets.forEach((sheet, i) => {
    const range = usedRange(sheet.cells);
    const size = range ? `${range.maxRow + 1}행 × ${range.maxCol + 1}열` : '빈 시트';
    const formulas = Object.values(sheet.cells).filter((c) => c.f).length;
    lines.push(`${i + 1}. **${sheet.name}** — ${size}, 수식 ${formulas}개`);
  });
  lines.push('');
  lines.push('---', '');

  sheets.forEach((sheet, i) => {
    const range = usedRange(sheet.cells);
    lines.push(`## 시트 ${i + 1} — ${sheet.name}`);
    lines.push('');

    if (!range) {
      lines.push('_빈 시트_', '', '---', '');
      return;
    }

    lines.push(`데이터 범위: \`A1:${toRef(range.maxCol, range.maxRow)}\``);
    lines.push('');

    const summary = summarizeColumns(sheet);
    if (summary.length) {
      const scope = summaryScope(sheet);
      lines.push('### 열 구성', '');
      if (scope) {
        const excluded = scope.excluded.length
          ? `, 집계 행 ${scope.excluded.map((r) => `${r}행`).join('/')}은 제외`
          : '';
        lines.push(`아래 통계는 ${scope.firstRow}–${scope.lastRow}행 데이터 기준입니다${excluded}.`);
        lines.push('');
      }
      for (const col of summary) {
        const stats =
          col.sum !== null
            ? ` (합계 ${col.sum}, 평균 ${col.avg}, 범위 ${col.min}~${col.max})`
            : '';
        lines.push(`- \`${col.col}\` **${col.header}** — ${col.type}, 값 ${col.count}개${stats}`);
      }
      lines.push('');
    }

    // Records read better than a grid when there is a header row.
    const headerRow = [];
    for (let c = 0; c <= range.maxCol; c++) {
      headerRow.push(displayValue(sheet.cells[toRef(c, 0)]) || indexToCol(c));
    }
    const hasHeader = headerRow.some((h, idx) => h !== indexToCol(idx));

    lines.push('### 데이터', '');
    const limit = Math.min(range.maxRow, GRID_DIGEST_ROWS);
    if (hasHeader) {
      for (let r = 1; r <= limit; r++) {
        const parts = [];
        for (let c = 0; c <= range.maxCol; c++) {
          const cell = sheet.cells[toRef(c, r)];
          const shown = displayValue(cell);
          if (shown === '') continue;
          parts.push(`${headerRow[c]}=${shown}`);
        }
        if (parts.length) lines.push(`- ${r + 1}행: ${parts.join(', ')}`);
      }
    } else {
      for (let r = 0; r <= limit; r++) {
        const parts = [];
        for (let c = 0; c <= range.maxCol; c++) {
          const shown = displayValue(sheet.cells[toRef(c, r)]);
          if (shown !== '') parts.push(`${toRef(c, r)}=${shown}`);
        }
        if (parts.length) lines.push(`- ${parts.join(', ')}`);
      }
    }
    if (range.maxRow > GRID_DIGEST_ROWS) {
      lines.push(`- _… ${range.maxRow - GRID_DIGEST_ROWS}개 행 생략_`);
    }
    lines.push('');

    const formulas = Object.entries(sheet.cells).filter(([, c]) => c.f);
    if (formulas.length) {
      lines.push('### 수식과 계산 결과', '');
      for (const [ref, cell] of formulas) {
        lines.push(`- \`${ref}\`: \`${cell.f}\` → **${displayValue(cell) || '(빈 값)'}**`);
      }
      lines.push('');
    }

    if (sheet.charts?.length) {
      lines.push('### 차트', '');
      for (const chart of sheet.charts) {
        const resolved = resolveChartSpec(chart.spec, sheet);
        lines.push(`- ${describeChart(resolved)}`);
        lines.push('');
        lines.push(chartToMarkdownTable(resolved));
        lines.push('');
      }
    }

    if (Object.keys(sheet.names ?? {}).length) {
      lines.push('### 이름 있는 범위', '');
      for (const [name, target] of Object.entries(sheet.names)) lines.push(`- \`${name}\` → \`${target}\``);
      lines.push('');
    }

    lines.push('---', '');
  });

  return `${lines.join('\n')}\n`;
}

function collapse(text) {
  return String(text).replace(/\s+/g, ' ').trim();
}

File: crates/ai-format/src/digest.rs
//! `AI.md` — a single markdown file that flattens a whole project for an LLM.
//!
//! The design bet: a model reasons far better about "상단 중앙, 전체 폭" than
//! about `x=96 y=220 w=1088`, and better about a sheet's column summary than
//! about 500 rows of numbers. So the digest is not a dump of the source files;
//! it is a translation of them into prose a model already understands, with
//! enough structure (slide numbers, cell addresses, heading levels) to stay
//! verifiable against the source.

use ai_formula::evaluate::display_value;
use ai_formula::refs::{index_to_col, to_ref};

use crate::blocks::Kind;
use crate::chart::{
    chart_to_markdown_table, describe_chart, parse_chart_block, resolve_spec, CellSource,
};
use crate::doc::describe_override;
use crate::geometry::{position_phrase, reading_order};
use crate::grid::{summarize_columns, summary_scope, used_range};
use crate::mdblocks::{count_words, heading_level, plain_text};
use crate::model::{Items, Project, Section, Sheet, Slide};

/// Build `AI.md` for a project.
pub fn build_digest(project: &Project) -> String {
    match &project.items {
        Items::Slides(slides) => deck_digest(project, slides),
        Items::Sections(sections) => doc_digest(project, sections),
        Items::Sheets(sheets) => grid_digest(project, sheets),
    }
}

fn header(project: &Project, subtitle: &str) -> Vec<String> {
    let m = &project.manifest;
    let modified = if m.modified.len() >= 10 {
        &m.modified[..10]
    } else {
        "—"
    };
    vec![
        format!("# {}", m.title),
        String::new(),
        format!("> {subtitle} · 최종 수정 {modified}"),
        String::new(),
        "<!-- 이 파일은 저장할 때마다 자동 생성됩니다. 직접 편집하면 다음 저장 시 덮어써집니다. -->"
            .to_string(),
        String::new(),
    ]
}

fn or_untitled(title: &str) -> &str {
    if title.is_empty() {
        "(제목 없음)"
    } else {
        title
    }
}

fn collapse(text: &str) -> String {
    text.split_whitespace().collect::<Vec<_>>().join(" ")
}

/* -------------------------------------------------------------------- deck */

fn deck_digest(project: &Project, slides: &[Slide]) -> String {
    let mut lines = header(
        project,
        &format!("AI Studio 프레젠테이션 · 슬라이드 {}장", slides.len()),
    );

    lines.push("## 목차".into());
    lines.push(String::new());
    for (i, slide) in slides.iter().enumerate() {
        lines.push(format!(
            "{}. {} — `{}`",
            i + 1,
            or_untitled(&slide.title),
            slide.layout_name
        ));
    }
    lines.push(String::new());
    lines.push("---".into());
    lines.push(String::new());

    for (i, slide) in slides.iter().enumerate() {
        lines.push(format!(
            "## 슬라이드 {} — {}",
            i + 1,
            or_untitled(&slide.title)
        ));
        lines.push(String::new());
        lines.push(format!("- 레이아웃: `{}`", slide.layout_name));
        lines.push(format!("- 요소 {}개", slide.blocks.len()));
        if !slide.notes.trim().is_empty() {
            lines.push(format!("- 발표자 노트: {}", collapse(&slide.notes)));
        }
        lines.push(String::new());

        let ordered = reading_order(&slide.blocks, |b| b.geometry(), 40.0);
        if ordered.is_empty() {
            lines.push("_빈 슬라이드_".into());
            lines.push(String::new());
            continue;
        }

        for (bi, block) in ordered.iter().enumerate() {
            let where_ = position_phrase(&block.geometry(), &slide.canvas);
            // A shape says which shape it is: "도형(판단)" carries the meaning of
            // a flowchart node, where a bare "도형" carries none.
            let kind = match (&block.kind, &block.shape) {
                (Kind::Shape, Some(shape)) => format!("{}({})", block.kind.label(), shape.label()),
                _ => block.kind.label().to_string(),
            };
            lines.push(format!("### {}) {kind} — {where_}", bi + 1));
            lines.push(String::new());
            let content = block.md.trim();

            // A chart is unreadable as a fenced JSON blob; give the model the
            // shape and the numbers instead.
            let chart = if block.kind == Kind::Chart {
                parse_chart_block(content)
            } else {
                None
            };
            match chart {
                Some(spec) => {
                    lines.push(describe_chart(&spec));
                    lines.push(String::new());
                    lines.push(chart_to_markdown_table(&spec));
                }
                None => lines.push(if content.is_empty() {
                    "_(빈 요소)_".to_string()
                } else {
                    content.to_string()
                }),
            }
            lines.push(String::new());
        }

        lines.push("---".into());
        lines.push(String::new());
    }

    let all_text = slides
        .iter()
        .flat_map(|s| s.blocks.iter().map(|b| plain_text(&b.md)))
        .filter(|t| !t.is_empty())
        .collect::<Vec<_>>()
        .join(" ");
    let element_count: usize = slides.iter().map(|s| s.blocks.len()).sum();
    let noted = slides.iter().filter(|s| !s.notes.trim().is_empty()).count();

    lines.push("## 전체 통계".into());
    lines.push(String::new());
    lines.push(format!("- 슬라이드: {}장", slides.len()));
    lines.push(format!("- 요소: {element_count}개"));
    lines.push(format!("- 본문 단어: 약 {}개", count_words(&all_text)));
    lines.push(format!("- 발표자 노트가 있는 슬라이드: {noted}장"));

    format!("{}\n", lines.join("\n"))
}

/* --------------------------------------------------------------------- doc */

fn doc_digest(project: &Project, sections: &[Section]) -> String {
    let mut lines = header(
        project,
        &format!("AI Studio 문서 · 섹션 {}개", sections.len()),
    );

    let outline: Vec<(usize, String)> = sections
        .iter()
        .flat_map(|s| s.blocks.iter())
        .filter_map(|b| {
            let level = heading_level(&b.md);
            (level > 0).then(|| (level, plain_text(&b.md)))
        })
        .collect();

    if !outline.is_empty() {
        lines.push("## 목차".into());
        lines.push(String::new());
        for (level, text) in &outline {
            lines.push(format!("{}- {text}", "  ".repeat(level.saturating_sub(1))));
        }
        lines.push(String::new());
        lines.push("---".into());
        lines.push(String::new());
    }

    for (si, section) in sections.iter().enumerate() {
        lines.push(format!("## 섹션 {} — {}", si + 1, section.name));
        lines.push(String::new());
        for block in &section.blocks {
            let md = block.md.trim();
            if md.is_empty() {
                continue;
            }
            match parse_chart_block(md) {
                Some(spec) => {
                    lines.push(describe_chart(&spec));
                    lines.push(String::new());
                    lines.push(chart_to_markdown_table(&spec));
                }
                None => lines.push(md.to_string()),
            }
            if let Some(o) = &block.format_override {
                lines.push(format!("<!-- 서식: {} -->", describe_override(o)));
            }
            lines.push(String::new());
        }
        lines.push("---".into());
        lines.push(String::new());
    }

    let text = sections
        .iter()
        .flat_map(|s| s.blocks.iter().map(|b| plain_text(&b.md)))
        .collect::<Vec<_>>()
        .join("\n");

    lines.push("## 전체 통계".into());
    lines.push(String::new());
    lines.push(format!("- 섹션: {}개", sections.len()));
    lines.push(format!("- 단어: 약 {}개", count_words(&text)));
    lines.push(format!(
        "- 글자(공백 제외): {}자",
        text.chars().filter(|c| !c.is_whitespace()).count()
    ));
    lines.push(format!("- 제목: {}개", outline.len()));

    format!("{}\n", lines.join("\n"))
}

/* -------------------------------------------------------------------- grid */

const GRID_DIGEST_ROWS: usize = 60;

fn grid_digest(project: &Project, sheets: &[Sheet]) -> String {
    // Recalculate here rather than trusting the caller's cached `v` values: the
    // digest is generated on save, and a caller may hand us cells whose formulas
    // have never been evaluated (an API client, or a freshly parsed file).
    // The whole workbook at once: a summary sheet's `=요약!B4` cannot be
    // resolved by recalculating its sheet alone.
    let sheets: Vec<Sheet> = crate::grid::recalculated_all(sheets);
    let mut lines = header(
        project,
        &format!("AI Studio 스프레드시트 · 시트 {}개", sheets.len()),
    );

    lines.push("## 시트 목록".into());
    lines.push(String::new());
    for (i, sheet) in sheets.iter().enumerate() {
        let size = match used_range(&sheet.cells) {
            Some(r) => format!("{}행 × {}열", r.max_row + 1, r.max_col + 1),
            None => "빈 시트".to_string(),
        };
        let formulas = sheet.cells.values().filter(|c| c.f.is_some()).count();
        lines.push(format!(
            "{}. **{}** — {size}, 수식 {formulas}개",
            i + 1,
            sheet.name
        ));
    }
    lines.push(String::new());
    lines.push("---".into());
    lines.push(String::new());

    for (i, sheet) in sheets.iter().enumerate() {
        lines.push(format!("## 시트 {} — {}", i + 1, sheet.name));
        lines.push(String::new());

        let Some(range) = used_range(&sheet.cells) else {
            lines.push("_빈 시트_".into());
            lines.push(String::new());
            lines.push("---".into());
            lines.push(String::new());
            continue;
        };

        lines.push(format!(
            "데이터 범위: `A1:{}`",
            to_ref(range.max_col, range.max_row)
        ));
        lines.push(String::new());

        let summary = summarize_columns(sheet);
        if !summary.is_empty() {
            lines.push("### 열 구성".into());
            lines.push(String::new());
            if let Some(scope) = summary_scope(sheet) {
                let excluded = if scope.excluded.is_empty() {
                    String::new()
                } else {
                    let rows: Vec<String> =
                        scope.excluded.iter().map(|r| format!("{r}행")).collect();
                    format!(", 집계 행 {}은 제외", rows.join("/"))
                };
                lines.push(format!(
                    "아래 통계는 {}–{}행 데이터 기준입니다{excluded}.",
                    scope.first_row, scope.last_row
                ));
                lines.push(String::new());
            }
            for col in &summary {
                let stats = match (&col.sum, &col.avg, &col.min, &col.max) {
                    (Some(sum), Some(avg), Some(min), Some(max)) => {
                        format!(" (합계 {sum}, 평균 {avg}, 범위 {min}~{max})")
                    }
                    _ => String::new(),
                };
                lines.push(format!(
                    "- `{}` **{}** — {}, 값 {}개{stats}",
                    col.col, col.header, col.col_type, col.count
                ));
            }
            lines.push(String::new());
        }

        // Records read better than a grid when there is a header row.
        let header_row: Vec<String> = (0..=range.max_col)
            .map(|c| {
                let shown = sheet
                    .cells
                    .get(&to_ref(c, 0))
                    .map(display_value)
                    .unwrap_or_default();
                if shown.is_empty() {
                    index_to_col(c)
                } else {
                    shown
                }
            })
            .collect();
        let has_header = header_row
            .iter()
            .enumerate()
            .any(|(idx, h)| *h != index_to_col(idx));

        lines.push("### 데이터".into());
        lines.push(String::new());
        let limit = range.max_row.min(GRID_DIGEST_ROWS);
        if has_header {
            for r in 1..=limit {
                let parts: Vec<String> = (0..=range.max_col)
                    .filter_map(|c| {
                        let shown = sheet
                            .cells
                            .get(&to_ref(c, r))
                            .map(display_value)
                            .unwrap_or_default();
                        (!shown.is_empty()).then(|| format!("{}={shown}", header_row[c]))
                    })
                    .collect();
                if !parts.is_empty() {
                    lines.push(format!("- {}행: {}", r + 1, parts.join(", ")));
                }
            }
        } else {
            for r in 0..=limit {
                let parts: Vec<String> = (0..=range.max_col)
                    .filter_map(|c| {
                        let shown = sheet
                            .cells
                            .get(&to_ref(c, r))
                            .map(display_value)
                            .unwrap_or_default();
                        (!shown.is_empty()).then(|| format!("{}={shown}", to_ref(c, r)))
                    })
                    .collect();
                if !parts.is_empty() {
                    lines.push(format!("- {}", parts.join(", ")));
                }
            }
        }
        if range.max_row > GRID_DIGEST_ROWS {
            lines.push(format!(
                "- _… {}개 행 생략_",
                range.max_row - GRID_DIGEST_ROWS
            ));
        }
        lines.push(String::new());

        let formulas: Vec<_> = sheet.cells.iter().filter(|(_, c)| c.f.is_some()).collect();
        if !formulas.is_empty() {
            lines.push("### 수식과 계산 결과".into());
            lines.push(String::new());
            for (reference, cell) in &formulas {
                let shown = display_value(cell);
                let shown = if shown.is_empty() {
                    "(빈 값)".to_string()
                } else {
                    shown
                };
                lines.push(format!(
                    "- `{reference}`: `{}` → **{shown}**",
                    cell.f.as_deref().unwrap_or("")
                ));
            }
            lines.push(String::new());
        }

        if !sheet.charts.is_empty() {
            lines.push("### 차트".into());
            lines.push(String::new());
            for chart in &sheet.charts {
                let resolved = resolve_spec(&chart.spec, Some(sheet as &dyn CellSource));
                lines.push(format!("- {}", describe_chart(&resolved)));
                lines.push(String::new());
                lines.push(chart_to_markdown_table(&resolved));
                lines.push(String::new());
            }
        }

        if !sheet.names.is_empty() {
            lines.push("### 이름 있는 범위".into());
            lines.push(String::new());
            for (name, target) in &sheet.names {
                let target = match target {
                    serde_json::Value::String(s) => s.clone(),
                    other => other.to_string(),
                };
                lines.push(format!("- `{name}` → `{target}`"));
            }
            lines.push(String::new());
        }

        lines.push("---".into());
        lines.push(String::new());
    }

    format!("{}\n", lines.join("\n"))
}
